src/persist/src/postgres.rs

```diff
@@ -339,39 +340,70 @@ impl Consensus for PostgresConsensus {
-            // - expected corresponds to the most recent sequence number
-            //   i.e. there is no other sequence number > expected already
-            //   present.
-            //
-            // This query has also been written to execute within a single
-            // network round-trip (instead of a slightly simpler implementation
-            // that would call `BEGIN` and have multiple `SELECT` queries).
-            let q = "INSERT INTO consensus SELECT $1, $2, $3 WHERE
-                     EXISTS (
-                        SELECT * FROM consensus WHERE shard = $1 AND sequence_number = $4
-                     )
-                     AND NOT EXISTS (
-                         SELECT * FROM consensus WHERE shard = $1 AND sequence_number > $4
-                     )
-                     ON CONFLICT DO NOTHING";
-            let client = self.get_connection().await?;
-            let statement = client.prepare_cached(q).await?;
-            client
-                .execute(
-                    &statement,
-                    &[&key, &new.seqno, &new.data.as_ref(), &expected],
+        if let Some(expected) = expected {
+            // This query is tuned for performance against CockroacbDB.
+            // The read for the latest sequence number should hit 1 row
+            // via reverse scan over the PK, and the insert is designed
+            // to hit the fast-path 1-phase commit of CRDB. Any changes
+            // to this query should ensure a  EXPLAIN ANALYZE (VERBOSE)
+            // query plan contains `auto commit` on the insert.
+            let cas_insert_query = r#"
+                WITH
+                    greatest_seqno AS (
+                        SELECT sequence_number FROM consensus
+                        WHERE shard = $1
+                        ORDER BY sequence_number DESC LIMIT 1
+                    )
+                    INSERT INTO consensus (shard, sequence_number, data)
+                    SELECT $1, $2, $3
+                    WHERE (SELECT * FROM greatest_seqno) = $4;
+            "#;
+            // We pipeline an additional read query to fetch the latest state.
+            // This helps reduce retries in situations where our client is
+            // already on older seqno before attempting its CaS.
+            let select_query = r#"
+                SELECT sequence_number, data FROM consensus
+                WHERE shard = $1
+                ORDER BY sequence_number DESC LIMIT 1
+            "#;
+            let insert_params: &[&(dyn ToSql + Sync)] =
+                &[&key, &new.seqno, &new.data.as_ref(), &expected];
+            let select_params: &[&(dyn ToSql + Sync)] = &[&key];
+
+            let (insert, select) = {
+                let client = self.get_connection().await?;
+                let insert = client.prepare_cached(cas_insert_query).await?;
+                let select = client.prepare_cached(select_query).await?;
+                join!(
```

Comment:
we could also use a `select!` or similar construct so we can early return on a successful insert, without needing to wait for the pipelined read to return. I don't think this is super critical, but could be nice if it's not too gnarly to write

Reply:
Assuming that deadpool supports pipelining it's pretty reasonable to do. Here's an example from the stash: https://github.com/MaterializeInc/materialize/blob/7f169e5214243ee3c96d3b2be55f62413faafbf7/src/stash/src/postgres.rs#L423

Reply:
This doesn't do what you want. This will issue the insert and select to cockroach and complete once both of them have returned. Cockroach will process these in order (it doesn't start work on the select until the insert is done). Pipelining only saves time if there were network roundtrips when you don't need the result of the first statement to compute a second statement.

If you want cockroach to process the select and insert in parallel, you need to use one connection per query.

Reply:
Mm that's a good point. I'm actually still not sure that's not what we want to do, because the longest part of our insert-or-select retry loop is the end-to-end latency of talking to Cockroach each time, so the less time spent waiting there the better. for now, going to punt on pipelined/parallel reads